### ISSUE TITLE
Fix citation display in UI so that citation preview appears when identifier is registered via Advanced Create

### ISSUE THREAD

**Issue by mariagould** [CDLUC3/ezid#407]
Summary of issue (November 2023): There is a bug in the EZID UI code that is preventing a DOI citation preview from being displayed when the identifier is created via the "Advanced" mode. Users appreciate the citation preview so that they can easily transmit this information to other sources. However, many users take advantage of the Advanced mode in order to provide a wider array of metadata. Therefore, the lack of citation preview in the Advanced route makes their work more cumbersome. This is a long-standing bug that we should address to improve the EZID user experience. 

To reproduce: In the EZID UI, create a DOI in "Simple" mode. A citation preview will be displayed on the resulting identifier landing page. Then create a DOI in "Advanced" mode. The citation preview will not be displayed. 

----
Background information:

**Librarian user feedback**: "One of the issues was that since I reserve a lot of DOIs in the UI in advance for researchers, it would save me some time if a formatted citation would appear on the EZID record after I create it. The Citation Preview does appear after creating a DOI in the simple form, but when using the advanced form (https://ezid.cdlib.org/create/advanced), which is what I use nearly 100% of the time, no citation preview gets displayed."

**Why this is important**: (1) _Make it easy for researchers to follow best practices for data citation_. Librarian wants to provide the full citation and DOI (instead of just the DOI) to the researcher to encourage the researcher to include the full citation in their publication in support of proper data citation. (2) _Save time_. Because of the bug, the librarian has to copy/paste the metadata by hand and reformat it to set up the citation for the researcher. If EZID can generate the citation automatically, this will save the librarian time. 

**Initial observation from Greg**: "The UI code is rendering a citation only if the identifier has individual fields (datacite.title, etc.), not if it has a DataCite XML record.  No real reason for that, it's a bug.  I can look into fixing it."

**Summary**: I believe that Greg never got around to fixing this so we should explore how to support this and fit it into our planning. 

Screenshot: Citation preview visible after creating a DOI via "simple" create
![image001.jpg](https://images.zenhubusercontent.com/5d7be665d6bc2d0001f77a8f/d63ed973-ca0f-48bd-9523-43373e31226e)

Screenshot: Citation preview not shown when creating a DOI via "advanced' create
![Screen Shot 2020-06-12 at 1.56.02 PM.png](https://images.zenhubusercontent.com/5d7be665d6bc2d0001f77a8f/c5b3d07a-0c6d-4f0b-b1eb-315fbd3b4511)

**Comment by mariagould** [comment]
@jsjiang just confirming that I have reproduced this again and can confirm the issue is persisting. This is ready to work on whenever you are available.

**Comment by jsjiang** [comment]
EZID UI only shows  the "Citation Preview" the metadata of the identifier meets the following two conditions:
1. metadata  profile is 'datacite', and
2. metadata does not has block data

[templates/manage/details.html](https://github.com/CDLUC3/ezid/blob/fd2868f2f9f48dbe0100186c214b75a9658e7c9a/templates/manage/details.html#L62C6-L79C14)

```
{% if not has_block_data and current_profile.name == 'datacite' %}
  <div class="row vertical-buffer-20">
    <div class="call-out">
      <h2 class="call-out__heading">{% trans "Citation Preview" %}</h2>
      <p class="truncated call-out__citation-indent">
        {% get_dict_value identifier "datacite.creator" %}
        ({% get_dict_value identifier "datacite.publicationyear" %}):
        {% get_dict_value identifier "datacite.title" %}.
        {% get_dict_value identifier "datacite.version" %}
        {% get_dict_value identifier "datacite.publisher" %}.
        {% if has_resource_type %}
          {% get_dict_value identifier "datacite.resourcetype" %}.
        {% endif %}
        {% active_id_display id_text testPrefixes %}
      </p>
    </div>
  </div>
  {% endif %}
```

The 'has_block_data' condition is determined by the profile and metadata:
1. profile is datacite, and
2. the dictionary type identifier's metadata has a key named as 'datacite'

So, the following datacite metadata does not have block data and is qualified for "Citation Preview" 
```
{
	"datacite.title": "test-title",
	"datacite.creator": "Jing-test-creator",
	"datacite.publisher": "test-publisher",
	"datacite.resourcetype": "Book",
	"datacite.publicationyear": "2023"
}
```

The following datacite metadata has block data and is not qualified for "Citation Preview" 
```
{
	"datacite": "<?xml version=\"1.0\"?>\n<resource xmlns=\"http://datacite.org/schema/kernel-4\" xmlns:xsi=\"http://www.w3.org/2001/XMLSchema-instance\" xsi:schemaLocation=\"http://datacite.org/schema/kernel-4 http://schema.datacite.org/meta/kernel-4/metadata.xsd\"><identifier identifierType=\"DOI\">10.24343/C35K5V</identifier><creators><creator><creatorName>Jing</creatorName></creator></creators><titles><title>test-title</title></titles><publisher>ACM</publisher><publicationYear>2023</publicationYear><resourceType resourceTypeGeneral=\"Audiovisual\">Audiovisual</resourceType></resource>"
}
```


[/impl/ui_common.py]
(https://github.com/CDLUC3/ezid/blob/fd2868f2f9f48dbe0100186c214b75a9658e7c9a/impl/ui_common.py#L234C1-L234C1)

```
def identifier_has_block_data(identifier):
    """Returns true if the identifier has block metadata, which affects both
    the display and the editability of the metadata in the UI."""
    return (identifier["_profile"] == "erc" and "erc" in identifier) or (
        identifier["_profile"] == "datacite" and "datacite" in identifier
    )
```


Sample identifiers and:
```
select identifier, crossrefStatus, metadata from ezidapp_identifier 
where identifier in  (
'doi:10.24343/C39884',  # datacite, with citation preview, datacite in datacite.xxxx format; 
'doi:10.24343/C35K5V',  # datacite, no citation, datacite in "datacite": xml; 
'doi:10.24343/C32S37',  # datacite, no citation, datacite in DC format 
'doi:10.15697/FK2363S', # crossref, no citation previdw
'ark:/99999/fk8ng4nwj'  # erc, no citation preview
);
```

**Comment by jsjiang** [comment]
details.html code for generating metadata from xml

```
  {% if datacite_html %}
    <style type="text/css">
			tr.dcms_identifier { display: none;}
    </style>
    {% autoescape off %}
      {{datacite_html}}
    {% endautoescape %}
    <p><a class="link__primary" href="/manage/display_xml/{{ id_text }}" target="_blank">{% trans "See the DataCite XML Metadata record in a new window" %}</a></p>
  {% endif %}
```

HTML produced from above code:

```
<style type="text/css">
--
  | tr.dcms_identifier { display: none;}
  | </style>
  |  
  | <table class="dcms_table"><tr class="dcms_element dcms_identifier"><th class="dcms_label dcms_identifier">Identifier:</th><td class="dcms_value dcms_identifier">doi:10.24343/C35K5V</td></tr><tr class="dcms_element dcms_creators"><th class="dcms_label dcms_creators">Creators:</th><td class="dcms_value dcms_creators">Jing</td></tr><tr class="dcms_element dcms_titles"><th class="dcms_label dcms_titles">Title:</th><td class="dcms_value dcms_titles">test-title</td></tr><tr class="dcms_element dcms_publisher"><th class="dcms_label dcms_publisher">Publisher:</th><td class="dcms_value dcms_publisher">ACM</td></tr><tr class="dcms_element dcms_publicationyear"><th class="dcms_label dcms_publicationyear">Publication year:</th><td class="dcms_value dcms_publicationyear">2023</td></tr><tr class="dcms_element dcms_resourcetype"><th class="dcms_label dcms_resourcetype">Resource type:</th><td class="dcms_value dcms_resourcetype">Audiovisual/Audiovisual</td></tr></table>
  |  
  | <p><a class="link__primary" href="/manage/display_xml/doi:10.24343/C35K5V" target="_blank">See the DataCite XML Metadata record in a new window</a></p>

<style type="text/css">
			tr.dcms_identifier { display: none;}
    </style>
    
      <table class="dcms_table"><tr class="dcms_element dcms_identifier"><th class="dcms_label dcms_identifier">Identifier:</th><td class="dcms_value dcms_identifier">doi:10.24343/C35K5V</td></tr><tr class="dcms_element dcms_creators"><th class="dcms_label dcms_creators">Creators:</th><td class="dcms_value dcms_creators">Jing</td></tr><tr class="dcms_element dcms_titles"><th class="dcms_label dcms_titles">Title:</th><td class="dcms_value dcms_titles">test-title</td></tr><tr class="dcms_element dcms_publisher"><th class="dcms_label dcms_publisher">Publisher:</th><td class="dcms_value dcms_publisher">ACM</td></tr><tr class="dcms_element dcms_publicationyear"><th class="dcms_label dcms_publicationyear">Publication year:</th><td class="dcms_value dcms_publicationyear">2023</td></tr><tr class="dcms_element dcms_resourcetype"><th class="dcms_label dcms_resourcetype">Resource type:</th><td class="dcms_value dcms_resourcetype">Audiovisual/Audiovisual</td></tr></table>
    
    <p><a class="link__primary" href="[/manage/display_xml/doi:10.24343/C35K5V](https://ezid-stg.cdlib.org/manage/display_xml/doi:10.24343/C35K5V)" target="_blank">See the DataCite XML Metadata record in a new window</a></p>

```

Code to produce datacite_html  from metadata:
```
if d['current_profile'].name == 'datacite' and 'datacite' in id_metadata:
        r = impl.datacite.dcmsRecordToHtml(id_metadata["datacite"])
        if r:
            d['datacite_html'] = r


...

def dcmsRecordToHtml(record):
    """Convert a DataCite Metadata Scheme <http://schema.datacite.org/> record
    to an XHTML table

    Args:
        record: The record should be unencoded.

    Returns:
        XHTML table or None on error.
    """
    try:
        r = lxml.etree.tostring(
            lxml.etree.XSLT(
                lxml.etree.parse(
                    os.path.join(django.conf.settings.PROJECT_ROOT, "profiles", "datacite.xsl")
                )
            )(impl.util.parseXmlString(record)),
            encoding=str,
        )
        assert r.startswith("<table")
        return r
    except Exception:
        return None
```

**Comment by jsjiang** [comment]
We may need another datacite stylesheet for extracting citation data elements from metadata in XML

**Comment by jsjiang** [comment]
Test IDs on ezid-stg:
* https://ezid-stg.cdlib.org/id/doi:10.24343/C35K5V - no Citation Preview
* https://ezid-stg.cdlib.org/id/doi:10.5072/FK2377GV8H - with Citation Preview

Test IDs on ezid-dev:
* http://uc3-ezidui01x2-dev.cdlib.org/id/doi:10.5072/FK2GQ71W88 - No Citation Preview (datacite with xml metadata)
* http://uc3-ezidui01x2-dev.cdlib.org/id/doi:10.5072/FK2MG7RQ2W - with Citation Preview (datacite with name/value format metadata)

**Comment by jsjiang** [comment]
@mariagould Hi Maria, can you provide following information/input:

Currently EZID shows the "Citation Preview" box when profile='Datacite' and metadata is in the name/value format:

datacite.creator: test creator
datacite.title: test title

I will add a fix to show the citation box when profile='Datacite'  and metadata is in XML format.

Do we have Datacite records with other metadata formats (erc, dc, etc.)? How to handle citation preview in these cases?

It looks like Datacite accepts multiple titles and creators in XML. How to display title and creator in this case? 

Can you provide some test records on the ezid-dev server with:
* multiple titles
* multiple author
* has version

Here is the code for the citation preview:
```
{% get_dict_value identifier "datacite.creator" %}
        ({% get_dict_value identifier "datacite.publicationyear" %}):
        {% get_dict_value identifier "datacite.title" %}.
        {% get_dict_value identifier "datacite.version" %}
        {% get_dict_value identifier "datacite.publisher" %}.
        {% if has_resource_type %}
          {% get_dict_value identifier "datacite.resourcetype" %}.
        {% endif %}
```

**Comment by mariagould** [comment]
Thanks @jsjiang 

To keep it as simple as possible for now:

- Exclude all other formats besides DataCite profile and XML
- If multiple titles, display primary title only
- If multiple authors, display first author only and add "et al." after the author name
- Exclude version

If we can get the above approach operational on stage, I will confirm with users that this makes sense or if other tweaks are needed.

**Comment by jsjiang** [comment]
Added `xmltodict` to the requirements files (requirements.txt and requirements-dev.txt):
```
+xmltodict~=0.13.0
```

Re-installed packages on ezid-dev:
```
pip install -r requirements.txt
```

ERROR: pip's dependency resolver does not currently take into account all the packages that are installed. This behaviour is the source of the following dependency conflicts.
ezid 3.0.0 requires Django==3.2.7, but you have django 3.2.23 which is incompatible.
ezid 3.0.0 requires feedparser==6.0.8, but you have feedparser 6.0.11 which is incompatible.
ezid 3.0.0 requires Jinja2==3.0.1, but you have jinja2 3.0.3 which is incompatible.
Successfully installed Django-3.2.23 Jinja2-3.0.3 MarkupSafe-2.1.3 Pygments-2.10.0 botocore-1.21.65 configparser-5.0.2 feedparser-6.0.11 freezegun-1.1.0 hjson-3.0.2 pytest-6.2.5 python-dateutil-2.8.2 s3transfer-0.5.2 setuptools-57.4.0

**Comment by jsjiang** [comment]
Fix is on ezid-dev for testing.

Jing tested on the following IDs:
* doi:10.5072/FK29Z9962V - simple Datacite format
* doi:10.5072/FK2GQ71W88 - XML with single title and author
* doi:10.5072/FK26978S2J - XML with multiple titles and authors

**Comment by jsjiang** [comment]
@mariagould Hi Maria, 
The fix is on ezid-dev (http://uc3-ezidui01x2-dev.cdlib.org/) for testing. Please test and let me know if you notice issues.
Thank you

Jing

**Comment by jsjiang** [comment]
Jing: add unit tests for new function `briefDataciteRecord(record)`

**Comment by jsjiang** [comment]
* Created test records on ezid-dev:
```
# simple datacite format
select * from ezidapp_identifier where identifier ='doi:10.5072/FK29Z9962V';

# one creator, one title without language code
select * from ezidapp_identifier where identifier ='doi:10.5072/FK2GQ71W88';

# one creator, one title with language code
select * from ezidapp_identifier where identifier ='doi:10.5072/FK2JH3PX1D';

# two creators, two titles with lang code
select * from ezidapp_identifier where identifier = 'doi:10.5072/FK26978S2J';

# one creator, two titles without lang code
select * from ezidapp_identifier where identifier ='doi:10.5072/FK2DR30377'
```

* Tested in UI on ezid-dev
* Created test records and added unit tests

**Comment by jsjiang** [comment]
Added additional handling for Datacite XML records with sub properties or `xml:lang` attribute:
* creatorName : nameType
* title : titleType
* publisher: `xml:lang` attribute

Created additional test records and added additional cases to unit test 
```
# two creators both with nameType, 2nd title with titletype, publisher with lang code
select * from ezidapp_identifier where identifier = 'doi:10.5072/FK2930ZP60';

# one creator with nameType; 2 titles both contain titleType
select * from ezidapp_identifier where identifier ='doi:10.5072/FK25B07W2P'
```

**Comment by jsjiang** [comment]
Issue with the online edit function: the creatorName field lost sub-property 'nameType' and publisher field lost the langue attribute after the record went through the online Edit process.

The following IDs were created use the same metadata. The 2nd one was edited online by adding the funder. The 'nameType' sub-property of the CreatorName and the  language attribute of the publisher field were lost.
* doi:10.5072/FK21J9J306: no online edit
* doi:10.5072/FK25B07W2P: online edited

Metadta:
```
_profile: datacite
_target: https://google.com
datacite: <?xml version="1.0" encoding="UTF-8"?><resource xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" xmlns="http://datacite.org/schema/kernel-4" xsi:schemaLocation="http://datacite.org/schema/kernel-4 https://schema.datacite.org/meta/kernel-4.4/metadata.xsd"><identifier identifierType="DOI">10.5072/example-full</identifier><creators><creator><creatorName nameType="Personal">test creator</creatorName><givenName>Elizabeth</givenName><familyName>Miller</familyName><nameIdentifier schemeURI="https://orcid.org/" nameIdentifierScheme="ORCID">0000-0001-5000-0007</nameIdentifier><affiliation>DataCite</affiliation></creator></creators><titles><title xml:lang="en-US" titleType="TranslatedTitle">test title</title><title xml:lang="en-US" titleType="Subtitle">Demonstration of DataCite Properties.</title></titles><publisher xml:lang="en">test publisher</publisher><publicationYear>1990</publicationYear><subjects><subject xml:lang="en-US" schemeURI="http://dewey.info/" subjectScheme="dewey" classificationCode="000">computer science</subject></subjects><language>en-US</language><resourceType resourceTypeGeneral="Dataset">Dataset</resourceType></resource>
```

**Comment by mariagould** [comment]
@jsjiang UCSD user has confirmed that everything looks good so this is ready for production deploy.

**Comment by jsjiang** [comment]
Deployed the fix with v3.2.3. Verified with the following records:

A test DataCite record in the Name/Value format
https://ezid.cdlib.org/id/doi:10.5072/FK2HX1HC5W

A DataCite record in XML
https://ezid.cdlib.org/id/doi:10.18739/A2Z892G9G

A DataCite record in XML with namespace ns0 prefix
https://ezid.cdlib.org/id/doi:10.17920/G96905
